src/hosting/cloudRunProxy.ts

```diff
@@ -0,0 +1,75 @@
+import { RequestHandler } from "express";
+import { get } from "lodash";
+
+import { errorRequestHandler, proxyRequestHandler } from "./proxy";
+import * as getProjectId from "../getProjectId";
+import * as logger from "../logger";
+import { cloudRunApiOrigin, request as apiRequest } from "../api";
+
+export interface CloudRunProxyOptions {
+  project?: string;
+}
+
+export interface CloudRunProxyRewrite {
+  run: {
+    serviceId: string;
+    region?: string;
+  };
+}
+
+const cloudRunCache: { [s: string]: string } = {};
+
+function getCloudRunUrl(rewrite: CloudRunProxyRewrite, projectId: string): Promise<string> {
+  const alreadyFetched = cloudRunCache[`${rewrite.run.region}/${rewrite.run.serviceId}`];
+  if (alreadyFetched) {
+    return Promise.resolve(alreadyFetched);
+  }
+
+  const path = `/v1alpha1/projects/${projectId}/locations/${rewrite.run.region ||
+    "us-central1"}/services/${rewrite.run.serviceId}`;
+  logger.info(`[hosting] Looking up Cloud Run service "${path}" for its URL`);
+  return apiRequest("GET", path, { origin: cloudRunApiOrigin, auth: true })
+    .then((res) => {
+      const url = get(res, "body.status.address.hostname");
+      if (!url) {
+        return Promise.reject("Cloud Run URL doesn't exist in response.");
+      }
+
+      cloudRunCache[`${rewrite.run.region}/${rewrite.run.serviceId}`] = url;
+      return url;
+    })
+    .catch((err) => {
+      const errInfo = `error looking up URL for Cloud Run service: ${err}`;
+      return Promise.reject(errInfo);
+    });
+}
+
+/**
+ * Returns a function which, given a CloudRunProxyRewrite, returns a Promise
+ * that resolves with a middleware-like function that proxies the request to
+ * the live Cloud Run service running within the given project.
+ */
+export default function(
+  options: CloudRunProxyOptions
+): (r: CloudRunProxyRewrite) => Promise<RequestHandler> {
+  return async (rewrite: CloudRunProxyRewrite) => {
+    if (!rewrite.run) {
```

Comment:
I'm not actually able to test this case because TypeScript won't allow it to happen (and I don't think it really can).  Should I just remove this check?

Reply:
I think having the error case is better than not. It should be safe to leave it.